pkg/backingstore/reconciler.go

```diff
@@ -795,3 +862,180 @@ func (r *Reconciler) ReconcilePool() error {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Reconciler) updatePodTemplate() error {
+	if r.Secret.StringData["AGENT_CONFIG"] == "" {
+		return util.NewPersistentError("No AgentConfig - Can't add Missing Agents",
+			fmt.Sprintf("Current Backing store spec is: %q", r.BackingStore.Spec.PVPool))
+	}
+	r.PodAgentTemplate.Spec.Containers[0].Env[1].ValueFrom = &corev1.EnvVarSource{
+		SecretKeyRef: &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: r.Secret.Name,
+			},
+			Key: "AGENT_CONFIG",
+		},
+	}
+	r.PodAgentTemplate.Spec.Containers[0].Image = r.NooBaa.Status.ActualImage
+	if r.NooBaa.Spec.ImagePullSecret == nil {
+		r.PodAgentTemplate.Spec.ImagePullSecrets =
+			[]corev1.LocalObjectReference{}
+	} else {
+		r.PodAgentTemplate.Spec.ImagePullSecrets =
+			[]corev1.LocalObjectReference{*r.NooBaa.Spec.ImagePullSecret}
+	}
+	r.PodAgentTemplate.Labels = map[string]string{"pool": r.BackingStore.Name}
+	return nil
+}
+
+func (r *Reconciler) updatePvcTemplate() error {
+	if r.BackingStore.Spec.PVPool.StorageClass != "" {
+		r.PvcAgentTemplate.Spec.StorageClassName = &r.BackingStore.Spec.PVPool.StorageClass
+	}
+	r.PvcAgentTemplate.Spec.Resources = *r.BackingStore.Spec.PVPool.VolumeResources
+	r.PvcAgentTemplate.Labels = map[string]string{"pool": r.BackingStore.Name}
+	return nil
+}
+
+func (r *Reconciler) deletePvPool() error {
+	podsList := &corev1.PodList{}
+	util.KubeList(podsList, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	for i := range podsList.Items {
+		pod := &podsList.Items[i]
+		util.RemoveFinalizer(pod, nbv1.Finalizer)
+		if !util.KubeUpdate(pod) {
+			r.Logger.Errorf("Pod %q failed to remove finalizer %q",
+				pod.Name, nbv1.Finalizer)
+		}
+	}
+	util.KubeDeleteAllOf(&corev1.Pod{}, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	util.KubeDeleteAllOf(&corev1.PersistentVolumeClaim{}, client.InNamespace(options.Namespace), client.MatchingLabels{"pool": r.BackingStore.Name})
+	return nil
+}
+
+func (r *Reconciler) upgradeBackingStore(sts *appsv1.StatefulSet) error {
+	r.Logger.Infof("Deleting old statefulset: %s", sts.Name)
+	agentConfig := sts.Spec.Template.Spec.Containers[0].Env[1].Value
+	replicas := sts.Spec.Replicas
+	stsName := sts.Name
+	util.KubeDelete(sts, client.PropagationPolicy("Orphan")) // delete STS leave pods behind
+	o := util.KubeObject(bundle.File_deploy_internal_secret_empty_yaml)
+	secret := o.(*corev1.Secret)
+	secret.Name = fmt.Sprintf("backing-store-%s-%s", nbv1.StoreTypePVPool, r.BackingStore.Name)
+	secret.Namespace = options.Namespace
+	util.KubeCheck(secret)
+	secret.StringData["AGENT_CONFIG"] = agentConfig // update secret for future pods
+	util.KubeUpdate(secret)
+	for i := 0; i < int(*replicas); i++ {
+		pod := &corev1.Pod{}
+		pod.Name = fmt.Sprintf("%s-%d", stsName, i)
+		pod.Namespace = r.BackingStore.Namespace
+		util.KubeCheck(pod)
```

Comment:
`replicas` can be different from the actual number of pods. check that the pod\PVC really exist

Reply:
Yeah I will put the kubeCheck in If... good catch